src/generators/adaptive-weights.js

```diff
@@ -1,195 +1,76 @@
-   */
-  indicesCircleTaxicab(center, radius) {
-    return circleTaxicab(radius)
-      .map(([dv, dh]) => {
-        const [v, h] = this.toVh(center);
-        /** @type { Point } */
-        const vh = [v + dv, h + dh];
-
-        return vh;
-      })
-      .filter((vh) => isInGrid([this.size, this.size], vh))
-      .map((vh) => this.fromVh(vh));
-  }
-
-  /**
-   * @returns { Point }
-   */
-  placeStone() {
-    const stone = pickIndex(this.weights);
-
-    //// points to exclude for future placements
-    this.weights[stone] = 0;
-
-    // within separation
-    range(1, this.separation).forEach((radius) => {
-      this.indicesCircleTaxicab(stone, radius).forEach((point) => {
-        this.weights[point] = 0;
-      });
-    });
-
-    //// points to increase weight
-    range(1 + this.separation, this.maxRadius(stone)).forEach((radius) => {
-      this.indicesCircleTaxicab(stone, radius).forEach((point) => {
-        this.weights[point] *= radius;
-      });
-    });
-
-    //// cut off peaks if too high
-    const weightMax = 2 * medianNonzero(this.weights);
-    this.weights = this.weights.map((wgt) => Math.min(wgt, weightMax));
-
-    return this.toVh(stone);
-  }
-}
+  distanceTaxicabMaker,
+} from "./utils/weights.js";
 
 /**
- * @param { number } numStone
+ * @param { number } totalStones
  * @param { Config } config
  * @returns { Placement[] }
  */
 export default function adaptiveWeights(
-  numStone,
+  totalStones,
   { size, margins, handicap, preventAdjacent },
 ) {
-  // TODO: Allow any separation.
   const separation = preventAdjacent ? 1 : 0;
-  const boardInner = new Board(size - 2 * margins, separation);
 
-  /** @type { Point[] } */
-  const stonesInner = [];
-
-  range(1, numStone).forEach(() => {
-    stonesInner.push(boardInner.placeStone());
-  });
-
-  const stonesOuter = stonesInner.map(([v, h]) => {
-    /** @type { Point } */
-    // '1 +' makes coordinates start at 1 instead of 0
-    const vhOuter = [1 + margins + v, 1 + margins + h];
+  /** @type { Point } */
+  const start = [margins, margins];
+  /** @type { Point } */
+  const end = [size - margins, size - margins];
 
-    return vhOuter;
-  });
+  const circleTaxicab = circleTaxicabMaker(start, end);
 
-  return assignPlayers(stonesOuter, handicap);
+  /** @type { Point[] } */
+  const stones = [];
+
+  range(0, totalStones).reduce(
+    (weights) => {
+      const stone = weights.toVh(pickIndex(weights.values));
+      stones.push(stone);
+
+      const distanceTaxicab = distanceTaxicabMaker(stone);
+
+      const weightsNew = weights
+        // exclude within separation
+        .applyAt(
+          () => 0,
+          range(0, separation + 1).flatMap((radius) =>
+            circleTaxicab(stone, radius),
+          ),
+        )
+        // multiply weight by taxicab distance
+        .apply(
+          /** @type { (wgt: number, idx: number) => number } */
+          (wgt, idx) => wgt * distanceTaxicab(weights.toVh(idx)),
+        );
```

Comment:
You need to change the annotation in `src/generators/utils.js:102` to accept an optional index.

```js
/**
 * @param { (val: number, ?idx: number) => number } func
 * @returns { Grid }
 */
```

Then you can remove this extra annotation.

Reply:
With `?`, type checker said "Identifier expected", underlining `?`.

However, without `?`, `(val: number, idx: number) => number` somehow worked.

Making the same change to `applyAt()` and `applyExcept()` did lead to type check errors, but so far they've only been used with `(val: number) => number` so I kept them as before.

Reply:
I guess I was wrong about the optional index. The question mark should go *after* the param name to be a correct syntax (not before as in my example). But this param isn’t optional at all. It is always provided in the `Array.prototype.map` method. In fact this array method also provides it self as the third param (which is not used in the `.apply` function). So the annotation could have been:

```js
/**
 * @param { (val: number, idx: number, arr: number[]) => number } func
 * @returns { Grid }
 */
```

and it would pass the typechecker.

Reply:
Yes, that also worked. Also, what was I saying, the other two already accept indices (in [vertical, horizontal]) as a separate parameter.